src/data-source-dos.cc

```diff
@@ -0,0 +1,566 @@
+/*
+ * data-source-dos.cc - DOS game resources file functions
+ *
+ * Copyright (C) 2014  Jon Lund Steffensen <jonlst@gmail.com>
+ *
+ * This file is part of freeserf.
+ *
+ * freeserf is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * freeserf is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with freeserf.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "src/data-source-dos.h"
+
+#include <cassert>
+#include <algorithm>
+#include <fstream>
+
+#include "src/misc.h"
+BEGIN_EXT_C
+  #include "src/freeserf_endian.h"
+  #include "src/log.h"
+END_EXT_C
+#include "src/tpwm.h"
+#include "src/data.h"
+#include "src/sfx2wav.h"
+#include "src/xmi2mid.h"
+
+#ifdef min
+# undef min
+#endif
+
+#ifdef max
+# undef max
+#endif
+
+/* There are different types of sprites:
+ - Non-packed, rectangular sprites: These are simple called sprites here.
+ - Transparent sprites, "transp": These are e.g. buldings/serfs.
+ The transparent regions are RLE encoded.
+ - Bitmap sprites: Conceptually these contain either 0 or 1 at each pixel.
+ This is used to either modify the alpha level of another sprite (shadows)
+ or mask parts of other sprites completely (mask sprites).
+ */
+
+data_source_dos_t::data_source_dos_t() {
+  sprites = NULL;
+  sprites_size = 0;
+  entry_count = 0;
+  animation_table = NULL;
+}
+
+data_source_dos_t::~data_source_dos_t() {
+  if (sprites != NULL) {
+    free(sprites);
+    sprites = NULL;
+  }
+
+  if (animation_table != NULL) {
+    delete[] animation_table;
+    animation_table = NULL;
+  }
+}
+
+bool
+data_source_dos_t::check(const std::string &path, std::string *load_path) {
+  const char *default_data_file[] = {
+    "SPAE.PA", /* English */
+    "SPAF.PA", /* French */
+    "SPAD.PA", /* German */
+    "SPAU.PA", /* Engish (US) */
+    NULL
+  };
+
+  for (const char **df = default_data_file; *df != NULL; df++) {
+    std::string file_path = path + '/' + *df;
+    LOGI("data", "Looking for game data in '%s'...", file_path.c_str());
+    if (check_file(file_path)) {
+      if (load_path != NULL) {
+        *load_path = file_path;
+      }
+      return true;
+    }
+  }
+
+  return false;
+}
+
+/* Load data file at path and let the global variable sprites refer to the memory
+ with the data file content. */
```

Comment:
This comment seems to be inaccurate now as there is no global variable `sprites`.